### PR TITLE
Replace base64_cipher.py with an easy to understand version

### PR DESCRIPTION
**Pull request by hfz1337**
### **Describe your change:**
This PR changes the existing `base64_cipher.py` with an easy to understand script that explains how the Base64 algorithm works under the hood. The solution is not necessarily faster, but I believe it will help people understand how the algorithm works and why it is possible to use Base64 in Steganography to conceal the existence of a piece of data.
Further more, I believe the `base64_cipher.py` name can be misleading as Base64 is considered an encoding, and not a cipher.

* [ ] Add an algorithm?
* [x] Improve an existing solution.
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Review by xcodz-dot — APPROVED** [review]

**Comment by xcodz-dot** [comment]
@cclauss, check this

File: ciphers/base64_encoding.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    >>> from base64 import b64encode
+    >>> a = b"This pull request is part of Hacktoberfest20!"
+    >>> b = b"https://tools.ietf.org/html/rfc4648"
+    >>> c = b"A"
+    >>> base64_encode(a) == b64encode(a)
+    True
+    >>> base64_encode(b) == b64encode(b)
+    True
+    >>> base64_encode(c) == b64encode(c)
+    True
```
Comment (cclauss) [inline ciphers/base64_encoding.py:29]: Please add a test that proves that this algorithm behaves like `b64encode("abc")`.

File: ciphers/base64_encoding.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    >>> from base64 import b64decode
+    >>> a = "VGhpcyBwdWxsIHJlcXVlc3QgaXMgcGFydCBvZiBIYWNrdG9iZXJmZXN0MjAh"
+    >>> b = "aHR0cHM6Ly90b29scy5pZXRmLm9yZy9odG1sL3JmYzQ2NDg="
+    >>> c = "QQ=="
+    >>> base64_decode(a) == b64decode(a)
+    True
+    >>> base64_decode(b) == b64decode(b)
+    True
+    >>> base64_decode(c) == b64decode(c)
+    True
```
Comment (cclauss) [inline ciphers/base64_encoding.py:73]: Please add a test that proves that this algorithm behaves like `b64decode("abc")`.

**Review by cclauss — CHANGES_REQUESTED** [review]
Please add these two tests to show how the algorithms deal with bad input data.

File: ciphers/base64_encoding.py [hunk 3]
```diff
@@ -27,8 +27,18 @@ def base64_encode(data: bytes) -> bytes:
...
+    >>> base64_encode("abc")
+    Traceback (most recent call last):
+      ...
+    TypeError: a bytes-like object is required, not 'str'
     """
-    binary_stream = "".join(bin(char)[2:].zfill(8) for char in data)
+    # Make sure the supplied data is a bytes-like object
+    if not isinstance(data, bytes):
+        raise TypeError(
+            "a bytes-like object is required, not '{}'".format(data.__class__.__name__)
```
Comment (cclauss) [inline ciphers/base64_encoding.py:38]:
```suggestion
            f"a bytes-like object is required, not '{data.__class__.__name__}'"
```

File: ciphers/base64_encoding.py [hunk 4]
```diff
@@ -71,7 +81,27 @@ def base64_decode(encoded_data: str) -> bytes:
...
+    >>> base64_decode("abc")
+    Traceback (most recent call last):
+      ...
+    AssertionError: Incorrect padding
     """
+    # Make sure encoded_data is either a string or a bytes-like object
+    if not isinstance(encoded_data, bytes) and not isinstance(encoded_data, str):
+        raise TypeError(
+            "argument should be a bytes-like object or ASCII string, not '{}'".format(
+                encoded_data.__class__.__name__
```
Comment (cclauss) [inline ciphers/base64_encoding.py:93]:
```suggestion
            f"argument should be a bytes-like object or ASCII string, not '{encoded_data.__class__.__name__}'"
```

File: ciphers/base64_encoding.py (hunk 2 above)
Comment (hfz1337) [reply ciphers/base64_encoding.py:83]: Thanks, I made the necessary changes, hopefully.
But there's something that I find weird in the official implementation:
```python
>>> from base64 import b64decode
>>> # this throws an exception
>>> b64decode("\xff")
Traceback (most recent call last):
  ...
ValueError: string argument should contain only ASCII characters
>>> # while this returns an empty bytes-like object
>>> b64decode(b"\xff")
b''
>>> b64decode("#")
b''
```
I contrast to my implementation where I made sure to throw an exception when needed.

shouldn't `b64decode("#")` for example throw an exception?

File: ciphers/base64_encoding.py (hunk 4 above)
Comment (hfz1337) [reply ciphers/base64_encoding.py:93]: I was using f-strings at the beginning, but doing so exceeds the 88 chars per line in this case, and flake8 reports an error.
What should I do?

File: ciphers/base64_encoding.py (hunk 2 above)
Comment (cclauss) [reply ciphers/base64_encoding.py:83]: Perhaps you should open an issue at https://bugs.python.org ;-)

File: ciphers/base64_encoding.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    >>> base64_decode("abc")
+    Traceback (most recent call last):
+      ...
+    AssertionError: Incorrect padding
+    """
+    # Make sure encoded_data is either a string or a bytes-like object
+    if not isinstance(encoded_data, bytes) and not isinstance(encoded_data, str):
+        raise TypeError(
+            "argument should be a bytes-like object or ASCII string, not '{}'".format(
+                encoded_data.__class__.__name__
```
Comment (cclauss) [inline ciphers/base64_encoding.py:93]:
```suggestion
            "argument should be a bytes-like object or ASCII string, not "
            f"'{encoded_data.__class__.__name__}'"
```

**Review by cclauss — APPROVED** [review]
An awesome contribution!  Thanks for doing this.

**Pull request merged**